File: pkg-r/tests/testthat/test-remote-dbsnp.R
# Offline tests for dbsnp remote mode. Existence and merge handling are checked
# against the NCBI RefSNP API; the biogate.remote_transport option replaces it.

.merged_body <- list(merged_snapshot_data = list(merged_into = list("7412")))

.stub_dbsnp <- function(status, body = NULL) {
  function(url, timeout) list(status = status, body = body)
}

test_that("the url strips the rs prefix", {
  resolver <- .remote_resolvers$dbsnp
  url <- resolver$url(NULL, "rs7412")
  expect_true(endsWith(url, "/refsnp/7412"))
})

test_that("a current rsID is valid", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biogate.remote_transport = .stub_dbsnp(200))
  res <- check_id("rs7412", "dbsnp", how = "remote")
  expect_true(res$valid)
  expect_identical(res$normalized, "rs7412")
  expect_true(is.na(res$suggestion))
})

test_that("a merged rsID is invalid and suggests the primary id", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())
  withr::local_options(
    biogate.remote_transport = .stub_dbsnp(200, .merged_body)
  )
  res <- check_id("rs3200542", "dbsnp", how = "remote")
  expect_false(res$valid)
  expect_identical(res$suggestion, "rs7412")
})

test_that("an absent rsID is invalid", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biogate.remote_transport = .stub_dbsnp(404))
  res <- check_id("rs999999999999", "dbsnp", how = "remote")
  expect_false(res$valid)
  expect_true(is.na(res$suggestion))
})

test_that("a malformed id skips the network", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())
  withr::local_options(
    biogate.remote_transport = function(url, timeout) {
      stop("a malformed id must not reach the network")
    }
  )
  res <- check_id("ss12345", "dbsnp", how = "remote")
  expect_false(res$valid)
  expect_true(is.na(res$suggestion))
})

test_that("an unexpected status raises and is not cached", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biogate.remote_transport = .stub_dbsnp(500))
  expect_error(
    check_id("rs7412", "dbsnp", how = "remote"),
    class = "biogate_error_remote"
  )
  expect_false(file.exists(.remote_cache_path("dbsnp", "refsnp", "rs7412")))
})

test_that("a merge round-trips through the cache", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())
  withr::local_options(
    biogate.remote_transport = .stub_dbsnp(200, .merged_body)
  )
  expect_identical(
    check_id("rs3200542", "dbsnp", how = "remote")$suggestion,
    "rs7412"
  )
  withr::local_options(
    biogate.remote_transport = function(url, timeout) {
      stop("network should not be called on a cache hit")
    }
  )
  expect_identical(
    check_id("rs3200542", "dbsnp", how = "remote")$suggestion,
    "rs7412"
  )
})

File: pkg-py/tests/test_remote_dbsnp.py
"""dbSNP remote mode: existence and merge handling via the NCBI RefSNP API."""

import pytest

import biogate
import biogate._remote as remote
from biogate import RemoteError

_MERGED = {"merged_snapshot_data": {"merged_into": ["7412"]}}


@pytest.fixture(autouse=True)
def _isolate_cache(tmp_path, monkeypatch):
    monkeypatch.setenv("BIOGATE_CACHE_DIR", str(tmp_path))


def _stub(status, body=None):
    def _http_get(url, timeout=30):
        return status, body

    return _http_get


def test_url_strips_the_rs_prefix():
    from biogate._registry import get_source

    url = remote._dbsnp_url(get_source("dbsnp"), "rs7412")
    assert url.endswith("/refsnp/7412")


def test_current_rsid_is_valid(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub(200))
    res = biogate.check_id("rs7412", source_db="dbsnp", how="remote")[0]
    assert res.valid is True
    assert res.normalized == "rs7412"
    assert res.suggestion is None


def test_merged_rsid_is_invalid_and_suggests_primary(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub(200, _MERGED))
    res = biogate.check_id("rs3200542", source_db="dbsnp", how="remote")[0]
    assert res.valid is False
    assert res.suggestion == "rs7412"


def test_absent_rsid_is_invalid(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub(404))
    res = biogate.check_id("rs999999999999", source_db="dbsnp", how="remote")[0]
    assert res.valid is False
    assert res.suggestion is None


def test_malformed_skips_the_network(monkeypatch):
    def _forbidden(url, timeout=30):
        raise AssertionError("a malformed id must not reach the network")

    monkeypatch.setattr(remote, "_http_get", _forbidden)
    res = biogate.check_id("ss12345", source_db="dbsnp", how="remote")[0]
    assert res.valid is False
    assert res.suggestion is None


def test_unexpected_status_raises_and_is_not_cached(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub(500))
    with pytest.raises(RemoteError):
        biogate.check_id("rs7412", source_db="dbsnp", how="remote")
    assert not remote._remote_cache_path("dbsnp", "refsnp", "rs7412").is_file()


def test_merge_round_trips_through_the_cache(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub(200, _MERGED))
    assert (
        biogate.check_id("rs3200542", source_db="dbsnp", how="remote")[0].suggestion
        == "rs7412"
    )

    def _forbidden(url, timeout=30):
        raise AssertionError("network should not be called on a cache hit")

    monkeypatch.setattr(remote, "_http_get", _forbidden)
    assert (
        biogate.check_id("rs3200542", source_db="dbsnp", how="remote")[0].suggestion
        == "rs7412"
    )
